internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,111 @@
+package logstate
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"github.com/canonical/pebble/internal/plan"
+)
+
+func init() {
+	RegisterLogBackend("loki", func(t *plan.LogTarget) (LogBackend, error) { return NewLokiBackend(t.Location) })
+}
+
+type LokiBackend struct {
+	address *url.URL
+}
+
+func NewLokiBackend(address string) (LogBackend, error) {
+	u, err := url.Parse(address)
+	if err != nil || u.Host == "" {
+		u, err = url.Parse("//" + address)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("invalid loki server address: %v", err)
+	} else if u.Scheme != "" && u.Scheme != "http" {
+		return nil, fmt.Errorf("unsupported loki address scheme '%v'", u.Scheme)
+	} else if u.RequestURI() != "" && u.RequestURI() != "/" {
+		return nil, fmt.Errorf("invalid loki address: extraneous path %q", u.RequestURI())
+	}
+
+	// check for and set loki defaults if ommitted from address
+	if u.Scheme == "" {
+		u.Scheme = "http"
+	}
+	if u.Port() == "" {
+		u.Host += ":3100"
+	}
+	u.Path = "/loki/api/v1/push"
+
+	b := &LokiBackend{address: u}
+	return b, nil
+}
+
+func (b *LokiBackend) Close() error { return nil }
+
+func (b *LokiBackend) Send(m *LogMessage) error {
+	data, err := json.Marshal(newLokiMessage(m))
+	if err != nil {
+		return fmt.Errorf("failed to build loki message: %v", err)
+	}
+
+	var buf bytes.Buffer
```

Comment:
probably excessive premature optimisation but would be good to avoid the added buffering and allow the json to be stream encoded.

```
type jsonGzipWriterTo struct {
  msg *lokiMessage
}

type writeCounter struct {
  n int64
  next io.Writer
}

func (w *writeCounter) Write(buf []byte) (n int, err error) {
  n, err = w.next.Write(buf)
  w.n+=int64(n)
  return
}

func (r *jsonGzipWriterTo) WriteTo(writer io.Writer) (n int64, err error) {
  counter := &writeCounter{next: writer}
  gz := gzip.NewWriter(counter)
  enc := json.NewEncoder(gz)
  err := enc.Encode(r.msg)
  return counter.n, err
}

func (r *jsonGzipWriterTo) Read(_ []byte) (int, error) {
  panic("unimplemented")
}
```